### DISCUSSION TITLE
Joomla content introtext/fulltext question

### DISCUSSION THREAD

**Discussion by woluweb** [joomlatools/joomlatools-pages#667] [Q&A] [answered]
Little question regarding introtext/fulltext. When I use `model: database?table=content` then `$article->introtext` works. But using `model: ext:joomla.model.articles` then `$article->introtext` is not working (not even in the dump)

NB: after playing I noticed that `$article->text` would show introtext if no readmore, fulltext only if readmore

**Comment by johanjanssens** [comment, marked as answer]
Happy to help.

You are using 2 different collection models that both get data from the Joomla content table, but each in a different way. 

## Database model

When you are using `model: database?table=content` you are accessing the Joomla content table directly. Pages will return all the rows and fields without any further processing. The Joomla content table has a `fulltext` and a `introtext` field that you can access

```
---
collection:
    model: database?table=content
---

<? foreach(collection() as $article) : ?>
<article>
   <?= $article->fulltext ?>
<article>
<? endforeach ?>
```

## Joomla extension

When you use `model: ext:joomla.model.articles`you are making use of the [Joomla extension](https://github.com/joomlatools/joomlatools-pages/pull/363) for Pages. This extension provided deeper integration with Joomla and offers an example how you can easily integrate Pages with any other Joomla extension. As a proof of concept II have also build an [extension for K2](https://github.com/joomlatools/joomlatools-pages/pull/365)

The `fulltext` and `introtext`properties don't exist when you use the Joomla extension. Those two go back 2 decades now, and it's time to get rid of them. For pages i'm moving to supporting **content modelling**, using **content types**. This means that properties for collection entities have the same name (if they are based on the content type) regardless of where the data is coming from, could be Joomla content, K2, a webservices ...

For example, both Pages, Joomla and the K2 integration all use the same base **article content type** and have the same entity properties and return the same type of data. This means that you can use Pages pages, Joomla content and K2 content with the same layouts and partials, you change the model type in the collection and everything just keeps working.

The reason I'm doing this is because it's makes layouts and partials re-usable. If a partial is build to use a specific content type you can use the partial with any collection that extends from that content type.

You can find a very basic definition a a content entity here: [/extensions/joomla/model/entity/abstract.php]( https://github.com/joomlatools/joomlatools-pages/blob/feature/175-joomla/contrib/extensions/joomla/model/entity/abstract.php)  This work is very much in progress, and quite rough but this is where I'm heading at.

### Some more resources on content modelling

- http://www.clevegibbon.com/content-modeling/elements-of-a-content-model/
- https://www.cmswire.com/content-strategy/content-modeling-what-it-is-and-how-to-get-started
- https://alistapart.com/article/content-modelling-a-master-skill/

### How is this different from the UCM that was introduced in Joomla 2.5?

1. The UCM enforces a content model through a fixed database schema. Content needs to be stored in the specific database tables for it to work and relations relay on database relations between tables. 

Pages is datasource agnostic and should work without a database, so I'm implementing the content model on a collection/entity (ORM layer) level, not a database level. This allows you to model any data, from any data source.

2.  The UCM enforces a single enforced approach, for Pages the unified content interface will not be enforced, you can structure your entity as you want, there will however be a base model entities you can extend from to provide a specific content type.

For example; both the article and category entity are implementing the abstract content entity:

- [/extensions/joomla/model/entity/article.php](https://github.com/joomlatools/joomlatools-pages/blob/feature/175-joomla/contrib/extensions/joomla/model/entity/article.php)
- [/extensions/joomla/model/entity/category.php](https://github.com/joomlatools/joomlatools-pages/blob/feature/175-joomla/contrib/extensions/joomla/model/entity/category.php)
